src/executor/executor.cpp

```diff
@@ -3,109 +3,222 @@
 
 #include "isa/instr.hpp"
 
+#include <bit>
+
 namespace sim {
 
 void Executor::exec(Instruction &instr, Hart &hart) const {
   m_exec_instr.at(instr.id)(instr, hart);
-  // Incorrect for branch instructions
-  hart.pc += 4;
 }
 
 void exec_ADD(Instruction &instr, Hart &hart) {
   hart.setReg(instr.rd, hart.getReg(instr.rs1) + hart.getReg(instr.rs2));
+  hart.pc += 4;
 } //!
 
 void exec_ADDI(Instruction &instr, Hart &hart) {
   hart.setReg(instr.rd, instr.imm + hart.getReg(instr.rs1));
+  hart.pc += 4;
 } //!
 
-void exec_AND(Instruction &instr, Hart &hart) {} //!
+void exec_AND(Instruction &instr, Hart &hart) {
+  hart.setReg(instr.rd, hart.getReg(instr.rs1) & hart.getReg(instr.rs2));
+  hart.pc += 4;
+} //!
 
-void exec_ANDI(Instruction &instr, Hart &hart) {} //!
+void exec_ANDI(Instruction &instr, Hart &hart) {
+  hart.setReg(instr.rd, hart.getReg(instr.rs1) & std::bit_cast<uint32_t>(instr.imm));
+  hart.pc += 4;
+} //!
 
-void exec_AUIPC(Instruction &instr, Hart &hart) {} //!
+void exec_AUIPC(Instruction &instr, Hart &hart) {
+  hart.setReg(instr.rd, instr.imm + hart.pc);
+  hart.pc += 4;
+} //!
 
-void exec_BEQ(Instruction &instr, Hart &hart) {} //!
+void exec_BEQ(Instruction &instr, Hart &hart) {
+  bool take = hart.getReg(instr.rs1) == hart.getReg(instr.rs2);
+  hart.pc += (take) ? instr.imm : 4;
+} //!
 
-void exec_BGE(Instruction &instr, Hart &hart) {} //!
+void exec_BGE(Instruction &instr, Hart &hart) {
+  bool take = std::bit_cast<int32_t>(hart.getReg(instr.rs1)) >= std::bit_cast<int32_t>(hart.getReg(instr.rs2));
+  hart.pc += (take) ? instr.imm : 4;
+} //!
 
-void exec_BGEU(Instruction &instr, Hart &hart) {} //!
+void exec_BGEU(Instruction &instr, Hart &hart) {
+  bool take = std::bit_cast<uint32_t>(hart.getReg(instr.rs1)) >= std::bit_cast<uint32_t>(hart.getReg(instr.rs2));
+  hart.pc += (take) ? instr.imm : 4;
+} //!
 
-void exec_BLT(Instruction &instr, Hart &hart) {} //!
+void exec_BLT(Instruction &instr, Hart &hart) {
+  bool take = std::bit_cast<int32_t>(hart.getReg(instr.rs1)) < std::bit_cast<int32_t>(hart.getReg(instr.rs2));
+  hart.pc += (take) ? instr.imm : 4;
+} //!
 
-void exec_BLTU(Instruction &instr, Hart &hart) {} //!
+void exec_BLTU(Instruction &instr, Hart &hart) {
+  bool take = std::bit_cast<uint32_t>(hart.getReg(instr.rs1)) < std::bit_cast<uint32_t>(hart.getReg(instr.rs2));
+  hart.pc += (take) ? instr.imm : 4;
+} //!
 
-void exec_BNE(Instruction &instr, Hart &hart) {} //!
+void exec_BNE(Instruction &instr, Hart &hart) {
+  bool take = hart.getReg(instr.rs1) != hart.getReg(instr.rs2);
+  hart.pc += (take) ? instr.imm : 4;
+} //!
 
-void exec_EBREAK(Instruction &instr, Hart &hart) {} //!
+void exec_EBREAK(Instruction &instr, Hart &hart) { hart.setTerminate(); } //!
 
 void exec_ECALL(Instruction &instr, Hart &hart) { hart.setTerminate(); } //!
 
-void exec_FENCE(Instruction &instr, Hart &hart) {} //!
+void exec_FENCE(Instruction &instr, Hart &hart) { hart.setTerminate(); } //!
```

Comment:
I would suggest just skipping FENCE instead of terminating execution. Since we don't have MMU yet, FENCE doesn't affect Hart state